### PR TITLE
GPTdiscord

### PR DESCRIPTION
**Pull request by USTAADCOM**

### CODE REVIEW

File: apps/GPTdiscord/run_discordGPT.py
Comment (maximiliangarmatsch) [inline apps/GPTdiscord/run_discordGPT.py:1]: file is way too big

File: apps/GPTdiscord/utils/helpers/solr_helper.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+import pysolr
+import os
+from colorama import Fore, Style
+import json
+
+SOLR_URL = "http://localhost:8983/solr/soupy"
```
Comment (maximiliangarmatsch) [inline apps/GPTdiscord/utils/helpers/solr_helper.py:6]: move url to config at top level or .env or something

File: apps/GPTdiscord/run_discordGPT.py
Comment (USTAADCOM) [reply apps/GPTdiscord/run_discordGPT.py:1]: make it less then 300 lines

File: apps/GPTdiscord/utils/helpers/solr_helper.py (hunk 1 above)
Comment (USTAADCOM) [reply apps/GPTdiscord/utils/helpers/solr_helper.py:6]: Done

File: README.md [hunk 2]
```diff
@@ -172,4 +172,59 @@ python start_discord.py
...
+!generate Image details
+```
+**Note**
+Image generation use dalle-3 which cost around $0.04 per image.  
+
+### 7. Refresh chat history
+
+```
+!refresh 
+```
```
Comment (maximiliangarmatsch) [inline README.md:229]: split into multiple files

File: apps/GPTdiscord/components/prompts/GPTbot_system_prompt.txt
Comment (maximiliangarmatsch) [inline apps/GPTdiscord/components/prompts/GPTbot_system_prompt.txt:1]: remove "_prompt" from filename

File: apps/GPTdiscord/components/prompts/GPTbot_user_prompt.txt
Comment (maximiliangarmatsch) [inline apps/GPTdiscord/components/prompts/GPTbot_user_prompt.txt:1]: delete

File: apps/GPTdiscord/run_discordGPT.py
Comment (maximiliangarmatsch) [inline apps/GPTdiscord/run_discordGPT.py:1]: file is definetly too big. split into 3-5 files please

**Review by maximiliangarmatsch — CHANGES_REQUESTED** [review]
Merge with pre-existing Discord Bot (whatever direction is easiest for you, but all functionality must remain)

File: apps/GPTdiscord/utils/helpers/image_analysis_helper.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+openai_api_key = os.getenv("OPENAI_API_KEY")
+
+
+async def encode_discord_image(image_url):
+    async with aiohttp.ClientSession() as session:
+        async with session.get(image_url) as response:
+            if response.status == 200:
+                image_data = await response.read()
+                image = Image.open(io.BytesIO(image_data)).convert("RGB")
+                if max(image.size) > 1000:
```
Comment (maximiliangarmatsch) [inline apps/GPTdiscord/utils/helpers/image_analysis_helper.py:19]: be more generous with line-breaks, it adds to readability

File: apps/GPTdiscord/utils/helpers/json_helper.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    if file_path.exists():
+        with open(file_path, "w") as file:
+            file.write("[]")
+        print(f"Cleared JSON data for channel ID: {channel.id}")
+        return True
+    else:
+        return False
+
+
+async def save_channel_history_to_json(channel):
```
Comment (maximiliangarmatsch) [inline apps/GPTdiscord/utils/helpers/json_helper.py:23]: can function be split into smaller?

File: apps/GPTdiscord/utils/helpers/json_helper.py
Comment (maximiliangarmatsch) [inline apps/GPTdiscord/utils/helpers/json_helper.py:1]: might it make sense to move general json_helpers into a shared util folder?

File: apps/GPTdiscord/utils/helpers/manage_history_helper.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    combined_results.extend(history)
+    return combined_results
+
+
+async def async_chat_completion(*args, **kwargs):
+    response = await asyncio.to_thread(openai.chat.completions.create, *args, **kwargs)
+    return response
+
+
+async def perform_tiered_solr_search(message_author, expanded_keywords):
```
Comment (maximiliangarmatsch) [inline apps/GPTdiscord/utils/helpers/manage_history_helper.py:68]: change naming

File: apps/GPTdiscord/utils/helpers/manage_history_helper.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+                combined_results.append(
+                    {"role": "user", "content": f"{username}: {content}", "tier": tier}
+                )
+            else:
+                print("Warning: Unexpected data structure in Solr result")
+    combined_results.extend(history)
+    return combined_results
+
+
+async def async_chat_completion(*args, **kwargs):
```
Comment (maximiliangarmatsch) [inline apps/GPTdiscord/utils/helpers/manage_history_helper.py:63]: "async" not neeeed

File: apps/GPTdiscord/utils/helpers/solr_helper.py
Comment (maximiliangarmatsch) [inline apps/GPTdiscord/utils/helpers/solr_helper.py:1]: rename or remove

File: README.md (hunk 2 above)
Comment (USTAADCOM) [reply README.md:229]: Seperate GPTdiscord setup Readme named Readme_GPTdiscord.md

File: apps/GPTdiscord/components/prompts/GPTbot_system_prompt.txt
Comment (USTAADCOM) [reply apps/GPTdiscord/components/prompts/GPTbot_system_prompt.txt:1]: Done

File: apps/GPTdiscord/components/prompts/GPTbot_user_prompt.txt
Comment (USTAADCOM) [reply apps/GPTdiscord/components/prompts/GPTbot_user_prompt.txt:1]: done

File: apps/GPTdiscord/utils/helpers/json_helper.py (hunk 4 above)
Comment (USTAADCOM) [reply apps/GPTdiscord/utils/helpers/json_helper.py:23]: Done

File: apps/GPTdiscord/utils/helpers/manage_history_helper.py (hunk 6 above)
Comment (USTAADCOM) [reply apps/GPTdiscord/utils/helpers/manage_history_helper.py:63]: Done

File: apps/GPTdiscord/utils/helpers/solr_helper.py
Comment (USTAADCOM) [reply apps/GPTdiscord/utils/helpers/solr_helper.py:1]: Done

File: apps/GPTdiscord/utils/helpers/manage_history_helper.py (hunk 5 above)
Comment (USTAADCOM) [reply apps/GPTdiscord/utils/helpers/manage_history_helper.py:68]: Done

File: apps/GPTdiscord/utils/helpers/image_analysis_helper.py (hunk 3 above)
Comment (USTAADCOM) [reply apps/GPTdiscord/utils/helpers/image_analysis_helper.py:19]: Done

File: apps/GPTdiscord/utils/helpers/json_helper.py
Comment (USTAADCOM) [reply apps/GPTdiscord/utils/helpers/json_helper.py:1]: Not yet will move all helpers in common general utils

File: apps/GPTdiscord/run_discordGPT.py
Comment (USTAADCOM) [reply apps/GPTdiscord/run_discordGPT.py:1]: Done

**Review by maximiliangarmatsch — APPROVED** [review]

**Pull request merged**